### PR TITLE
Introduced EntryMergedListener for listening WAN Replication events on t...

### LINKED ISSUE #2981: EntryListeners not called when entries modified by WAN replication
Because updates are sent to the remote cluster as MergeOperations, EntryListeners on the remote side are not being called. There either needs to be an entryMerged() method on EntryListener or the appropriate entryAdded/Removed/Updated method called when an update from the WAN is applied.

### PR DESCRIPTION
**Pull request by eminn**
...he receiving cluster.

This pull request is rebased and refactored version of @dbrimley 's #4122 .

Fixes #2981

### CODE REVIEW

**Comment by mahileeb** [comment]
Hooray! Thanks guys

File: hazelcast/src/main/java/com/hazelcast/map/impl/EntryListenerAdaptors.java [hunk 1]
```diff
@@ -166,7 +173,8 @@ private EntryListenerAdaptors() {
      * @return an array of {@link com.hazelcast.map.impl.ListenerAdapter}
      */
     public static ListenerAdapter[] createListenerAdapters(EntryListener listener) {
-        EntryEventType[] values = EntryEventType.values();
+        // We only care about these reference event types for backward compatibility.
+        EntryEventType[] values = new EntryEventType[] { ADDED, REMOVED, EVICTED , UPDATED , EVICT_ALL, CLEAR_ALL};
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/EntryListenerAdaptors.java]: :+1:

File: hazelcast/src/test/java/com/hazelcast/wan/WanReplicationTest.java [hunk 2]
```diff
@@ -533,6 +534,28 @@ public void replicationRing() {
...
 
+    @Test
+    public void willFireNewOnMergeEventAtReceivingCluster() {
+        setupReplicateFrom(configA, configB, clusterB.length, "atob", PassThroughMergePolicy.class.getName());
+
+        initClusterA();
+        initClusterB();
+
+        IMap<String, String> map = clusterB[0].getMap("map");
+        final CountDownLatch mergeEventFiredCounter = new CountDownLatch(100);
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/wan/WanReplicationTest.java]: seems this should be 1000 instead of 100?

**Comment by ahmetmircik** [comment]
Seems mergeable to me when it is green.

**Comment by gurbuzali** [comment]
needs rebase

**Comment by gurbuzali** [comment]
will merge once green

**Pull request merged**